File: lib-database-entities/src/main/java/com/ustadmobile/lib/db/entities/StateContentEntity.java
package com.ustadmobile.lib.db.entities;

import com.ustadmobile.lib.database.annotation.UmEntity;
import com.ustadmobile.lib.database.annotation.UmPrimaryKey;
import com.ustadmobile.lib.database.annotation.UmSyncLastChangedBy;
import com.ustadmobile.lib.database.annotation.UmSyncLocalChangeSeqNum;
import com.ustadmobile.lib.database.annotation.UmSyncMasterChangeSeqNum;

import static com.ustadmobile.lib.db.entities.StateContentEntity.TABLE_ID;

@UmEntity(tableId = TABLE_ID)
public class StateContentEntity {

    public static final int TABLE_ID = 72;

    @UmPrimaryKey(autoGenerateSyncable = true)
    private long stateContentUid;

    private long stateContentStateUid;

    private String stateContentKey;

    private String stateContentValue;

    private boolean isactive;

    @UmSyncMasterChangeSeqNum
    private long stateContentMasterChangeSeqNum;

    @UmSyncLocalChangeSeqNum
    private long stateContentLocalChangeSeqNum;

    @UmSyncLastChangedBy
    private int stateContentLastChangedBy;

    public StateContentEntity(String key, long stateUid, String valueOf, boolean isActive) {
        this.stateContentKey = key;
        this.stateContentValue = valueOf;
        this.isactive = isActive;
        this.stateContentStateUid = stateUid;
    }

    public StateContentEntity(){

    }

    public long getStateContentUid() {
        return stateContentUid;
    }

    public void setStateContentUid(long stateContentUid) {
        this.stateContentUid = stateContentUid;
    }

    public long getStateContentStateUid() {
        return stateContentStateUid;
    }

    public void setStateContentStateUid(long stateContentStateUid) {
        this.stateContentStateUid = stateContentStateUid;
    }

    public String getStateContentKey() {
        return stateContentKey;
    }

    public void setStateContentKey(String stateContentKey) {
        this.stateContentKey = stateContentKey;
    }

    public String getStateContentValue() {
        return stateContentValue;
    }

    public void setStateContentValue(String stateContentValue) {
        this.stateContentValue = stateContentValue;
    }

    public long getStateContentMasterChangeSeqNum() {
        return stateContentMasterChangeSeqNum;
    }

    public void setStateContentMasterChangeSeqNum(long stateContentMasterChangeSeqNum) {
        this.stateContentMasterChangeSeqNum = stateContentMasterChangeSeqNum;
    }

    public long getStateContentLocalChangeSeqNum() {
        return stateContentLocalChangeSeqNum;
    }

    public void setStateContentLocalChangeSeqNum(long stateContentLocalChangeSeqNum) {
        this.stateContentLocalChangeSeqNum = stateContentLocalChangeSeqNum;
    }

    public int getStateContentLastChangedBy() {
        return stateContentLastChangedBy;
    }

    public void setStateContentLastChangedBy(int stateContentLastChangedBy) {
        this.stateContentLastChangedBy = stateContentLastChangedBy;
    }

    public boolean isIsactive() {
        return isactive;
    }

    public void setIsactive(boolean isactive) {
        this.isactive = isactive;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;

        StateContentEntity that = (StateContentEntity) o;

        if (stateContentUid != that.stateContentUid) return false;
        if (stateContentStateUid != that.stateContentStateUid) return false;
        if (isactive != that.isactive) return false;
        if (stateContentKey != null ? !stateContentKey.equals(that.stateContentKey) : that.stateContentKey != null)
            return false;
        return stateContentValue != null ? stateContentValue.equals(that.stateContentValue) : that.stateContentValue == null;
    }

    @Override
    public int hashCode() {
        int result = (int) (stateContentUid ^ (stateContentUid >>> 32));
        result = 31 * result + (int) (stateContentStateUid ^ (stateContentStateUid >>> 32));
        result = 31 * result + (stateContentKey != null ? stateContentKey.hashCode() : 0);
        result = 31 * result + (stateContentValue != null ? stateContentValue.hashCode() : 0);
        result = 31 * result + (isactive ? 1 : 0);
        return result;
    }
}

File: lib-database-entities/src/commonMain/kotlin/com/ustadmobile/lib/db/entities/StateContentEntity.kt
package com.ustadmobile.lib.db.entities

import androidx.room.Entity
import androidx.room.PrimaryKey
import com.ustadmobile.door.annotation.*
import com.ustadmobile.lib.db.entities.StateContentEntity.Companion.TABLE_ID
import kotlinx.serialization.Serializable

@Entity
@Serializable
@ReplicateEntity(tableId = StateContentEntity.TABLE_ID, tracker = StateContentEntityReplicate::class)
@Triggers(arrayOf(
 Trigger(
     name = "statecontententity_remote_insert",
     order = Trigger.Order.INSTEAD_OF,
     on = Trigger.On.RECEIVEVIEW,
     events = [Trigger.Event.INSERT],
     sqlStatements = [
         """REPLACE INTO StateContentEntity(stateContentUid, stateContentStateUid, stateContentKey, stateContentValue, isIsactive, stateContentMasterChangeSeqNum, stateContentLocalChangeSeqNum, stateContentLastChangedBy, stateContentLct) 
         VALUES (NEW.stateContentUid, NEW.stateContentStateUid, NEW.stateContentKey, NEW.stateContentValue, NEW.isIsactive, NEW.stateContentMasterChangeSeqNum, NEW.stateContentLocalChangeSeqNum, NEW.stateContentLastChangedBy, NEW.stateContentLct) 
         /*psql ON CONFLICT (stateContentUid) DO UPDATE 
         SET stateContentStateUid = EXCLUDED.stateContentStateUid, stateContentKey = EXCLUDED.stateContentKey, stateContentValue = EXCLUDED.stateContentValue, isIsactive = EXCLUDED.isIsactive, stateContentMasterChangeSeqNum = EXCLUDED.stateContentMasterChangeSeqNum, stateContentLocalChangeSeqNum = EXCLUDED.stateContentLocalChangeSeqNum, stateContentLastChangedBy = EXCLUDED.stateContentLastChangedBy, stateContentLct = EXCLUDED.stateContentLct
         */"""
     ]
 )
))
//TODO
class StateContentEntity {

    @PrimaryKey(autoGenerate = true)
    var stateContentUid: Long = 0

    var stateContentStateUid: Long = 0

    var stateContentKey: String? = null

    var stateContentValue: String? = null

    var isIsactive: Boolean = false

    @MasterChangeSeqNum
    var stateContentMasterChangeSeqNum: Long = 0

    @LocalChangeSeqNum
    var stateContentLocalChangeSeqNum: Long = 0

    @LastChangedBy
    var stateContentLastChangedBy: Int = 0

    @LastChangedTime
    @ReplicationVersionId
    var stateContentLct: Long = 0

    constructor(key: String, stateUid: Long, valueOf: String, isActive: Boolean) {
        this.stateContentKey = key
        this.stateContentValue = valueOf
        this.stateContentStateUid = stateUid
        this.isIsactive = isActive
    }

    constructor() {

    }

    override fun equals(o: Any?): Boolean {
        if (this === o) return true
        if (o == null || this::class != o::class) return false

        val that = o as StateContentEntity?

        if (stateContentUid != that!!.stateContentUid) return false
        if (stateContentStateUid != that.stateContentStateUid) return false
        if (isIsactive != that.isIsactive) return false
        return if (stateContentKey != that.stateContentKey) false else stateContentValue == that.stateContentValue
    }

    override fun hashCode(): Int {
        var result = (stateContentUid xor stateContentUid.ushr(32)).toInt()
        result = 31 * result + (stateContentStateUid xor stateContentStateUid.ushr(32)).toInt()
        result = 31 * result + if (stateContentKey != null) stateContentKey!!.hashCode() else 0
        result = 31 * result + if (stateContentValue != null) stateContentValue!!.hashCode() else 0
        result = 31 * result + if (isIsactive) 1 else 0
        return result
    }

    companion object {

        const val TABLE_ID = 72
    }
}